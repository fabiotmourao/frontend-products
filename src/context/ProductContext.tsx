import React, { createContext, useContext, useEffect, useState } from 'react';
import { Product } from '../types/product';
import { getMockProducts } from '../services/api';

type ProductContextType = {
  products: Product[];
  setProducts: React.Dispatch<React.SetStateAction<Product[]>>;
};

export const ProductContext = createContext<ProductContextType | undefined>(undefined);

export const ProductProvider = ({ children }: { children: React.ReactNode }) => {
  const [products, setProducts] = useState<Product[]>([]);

  useEffect(() => {
    getMockProducts().then(setProducts);
  }, []);

  return (
    <ProductContext.Provider value={{ products, setProducts }}>
      {children}
    </ProductContext.Provider>
  );
};

export const useProductContext = () => {
  const context = useContext(ProductContext);
  if (!context) {
    throw new Error('useProductContext must be used within a ProductProvider');
  }
  return context;
};