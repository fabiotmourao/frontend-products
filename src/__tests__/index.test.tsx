import React from 'react';
import { render, act } from '@testing-library/react';
import Home from '../pages/index';
import { ProductProvider } from '../context/ProductContext';

describe('Snapshot da tela Home', () => {
  it('deve renderizar e bater com o snapshot', async () => {
    let asFragment: () => DocumentFragment = () => document.createDocumentFragment(); // Initialize as empty DocumentFragment
    await act(async () => {
      const result = render(
        <ProductProvider>
          <Home />
        </ProductProvider>
      );
      asFragment = result.asFragment;
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
