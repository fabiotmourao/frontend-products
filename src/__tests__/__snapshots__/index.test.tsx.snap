// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot da tela Home deve renderizar e bater com o snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-6"
  >
    <div
      class="flex justify-between items-center mb-6"
    >
      <h1
        class="text-2xl font-bold"
      >
        Gerenciamento de Produtos
      </h1>
      <button
        class="px-4 py-2 bg-blue-600 text-white rounded hover:bg-blue-700 transition"
      >
        Cadastrar Produto
      </button>
    </div>
    <div
      class="flex flex-wrap gap-4 items-end mb-6 bg-white p-4 rounded shadow"
    >
      <div
        class="flex flex-col"
      >
        <label
          class="text-sm font-medium"
        >
          Buscar por nome
        </label>
        <input
          class="border border-gray-300 rounded px-3 py-2"
          placeholder="Ex: Monitor"
          type="text"
          value=""
        />
      </div>
      <div
        class="flex flex-col"
      >
        <label
          class="text-sm font-medium"
        >
          Preço mínimo
        </label>
        <input
          class="border border-gray-300 rounded px-3 py-2"
          placeholder="0.00"
          type="number"
          value=""
        />
      </div>
      <div
        class="flex flex-col"
      >
        <label
          class="text-sm font-medium"
        >
          Preço máximo
        </label>
        <input
          class="border border-gray-300 rounded px-3 py-2"
          placeholder="9999.99"
          type="number"
          value=""
        />
      </div>
    </div>
    <div
      class="overflow-x-auto bg-white shadow rounded-lg"
    >
      <table
        class="min-w-full text-sm text-left text-gray-700"
      >
        <thead
          class="text-xs text-gray-700 uppercase bg-gray-100"
        >
          <tr>
            <th
              class="px-4 py-3"
            >
              Imagem
            </th>
            <th
              class="px-4 py-3 cursor-pointer"
            >
              Produto ⇅
            </th>
            <th
              class="px-4 py-3 cursor-pointer"
            >
              Categoria ⇅
            </th>
            <th
              class="px-4 py-3 cursor-pointer"
            >
              Preço ⇅
            </th>
            <th
              class="px-4 py-3 cursor-pointer"
            >
              Descrição ⇅
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="border-t hover:bg-gray-50 transition-all"
          >
            <td
              class="px-4 py-2"
            >
              <img
                alt="Monitor 24""
                class="rounded object-cover"
                data-nimg="1"
                decoding="async"
                height="48"
                loading="lazy"
                src="/_next/image?url=%2Fimages%2Fmonitor.png&w=96&q=75"
                srcset="/_next/image?url=%2Fimages%2Fmonitor.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fmonitor.png&w=96&q=75 2x"
                style="color: transparent;"
                width="48"
              />
            </td>
            <td
              class="px-4 py-2 font-semibold text-gray-800"
            >
              Monitor 24"
            </td>
            <td
              class="px-4 py-2"
            >
              Eletrônico
            </td>
            <td
              class="px-4 py-2 font-medium text-gray-900"
            >
              R$ 899.90
            </td>
            <td
              class="px-4 py-2 text-gray-600 max-w-[200px] truncate"
            >
              Monitor Full HD com painel IPS
            </td>
          </tr>
          <tr
            class="border-t hover:bg-gray-50 transition-all"
          >
            <td
              class="px-4 py-2"
            >
              <img
                alt="Teclado Mecânico"
                class="rounded object-cover"
                data-nimg="1"
                decoding="async"
                height="48"
                loading="lazy"
                src="/_next/image?url=%2Fimages%2Fteclado.png&w=96&q=75"
                srcset="/_next/image?url=%2Fimages%2Fteclado.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fteclado.png&w=96&q=75 2x"
                style="color: transparent;"
                width="48"
              />
            </td>
            <td
              class="px-4 py-2 font-semibold text-gray-800"
            >
              Teclado Mecânico
            </td>
            <td
              class="px-4 py-2"
            >
              Acessórios
            </td>
            <td
              class="px-4 py-2 font-medium text-gray-900"
            >
              R$ 299.99
            </td>
            <td
              class="px-4 py-2 text-gray-600 max-w-[200px] truncate"
            >
              Teclado com switches blue e iluminação RGB
            </td>
          </tr>
          <tr
            class="border-t hover:bg-gray-50 transition-all"
          >
            <td
              class="px-4 py-2"
            >
              <img
                alt="Mouse sem Fio"
                class="rounded object-cover"
                data-nimg="1"
                decoding="async"
                height="48"
                loading="lazy"
                src="/_next/image?url=%2Fimages%2Fmouse.png&w=96&q=75"
                srcset="/_next/image?url=%2Fimages%2Fmouse.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fmouse.png&w=96&q=75 2x"
                style="color: transparent;"
                width="48"
              />
            </td>
            <td
              class="px-4 py-2 font-semibold text-gray-800"
            >
              Mouse sem Fio
            </td>
            <td
              class="px-4 py-2"
            >
              Acessórios
            </td>
            <td
              class="px-4 py-2 font-medium text-gray-900"
            >
              R$ 49.99
            </td>
            <td
              class="px-4 py-2 text-gray-600 max-w-[200px] truncate"
            >
              Mouse sem fio com iluminação RGB
            </td>
          </tr>
          <tr
            class="border-t hover:bg-gray-50 transition-all"
          >
            <td
              class="px-4 py-2"
            >
              <img
                alt="Cadeira Gamer"
                class="rounded object-cover"
                data-nimg="1"
                decoding="async"
                height="48"
                loading="lazy"
                src="/_next/image?url=%2Fimages%2Fcadeira.png&w=96&q=75"
                srcset="/_next/image?url=%2Fimages%2Fcadeira.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fcadeira.png&w=96&q=75 2x"
                style="color: transparent;"
                width="48"
              />
            </td>
            <td
              class="px-4 py-2 font-semibold text-gray-800"
            >
              Cadeira Gamer
            </td>
            <td
              class="px-4 py-2"
            >
              Acessórios
            </td>
            <td
              class="px-4 py-2 font-medium text-gray-900"
            >
              R$ 599.99
            </td>
            <td
              class="px-4 py-2 text-gray-600 max-w-[200px] truncate"
            >
              Cadeira gamer com ajuste de altura e iluminação RGB
            </td>
          </tr>
          <tr
            class="border-t hover:bg-gray-50 transition-all"
          >
            <td
              class="px-4 py-2"
            >
              <img
                alt="Produto 5"
                class="rounded object-cover"
                data-nimg="1"
                decoding="async"
                height="48"
                loading="lazy"
                src="/_next/image?url=%2Fimages%2Fproduto5.png&w=96&q=75"
                srcset="/_next/image?url=%2Fimages%2Fproduto5.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fproduto5.png&w=96&q=75 2x"
                style="color: transparent;"
                width="48"
              />
            </td>
            <td
              class="px-4 py-2 font-semibold text-gray-800"
            >
              Produto 5
            </td>
            <td
              class="px-4 py-2"
            >
              Categoria 5
            </td>
            <td
              class="px-4 py-2 font-medium text-gray-900"
            >
              R$ 500.00
            </td>
            <td
              class="px-4 py-2 text-gray-600 max-w-[200px] truncate"
            >
              Descrição do Produto 5
            </td>
          </tr>
          <tr
            class="border-t hover:bg-gray-50 transition-all"
          >
            <td
              class="px-4 py-2"
            >
              <img
                alt="Produto 6"
                class="rounded object-cover"
                data-nimg="1"
                decoding="async"
                height="48"
                loading="lazy"
                src="/_next/image?url=%2Fimages%2Fproduto6.png&w=96&q=75"
                srcset="/_next/image?url=%2Fimages%2Fproduto6.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fproduto6.png&w=96&q=75 2x"
                style="color: transparent;"
                width="48"
              />
            </td>
            <td
              class="px-4 py-2 font-semibold text-gray-800"
            >
              Produto 6
            </td>
            <td
              class="px-4 py-2"
            >
              Categoria 6
            </td>
            <td
              class="px-4 py-2 font-medium text-gray-900"
            >
              R$ 600.00
            </td>
            <td
              class="px-4 py-2 text-gray-600 max-w-[200px] truncate"
            >
              Descrição do Produto 6
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="flex justify-center mt-6"
    >
      <nav
        class="flex items-center space-x-1 text-sm text-gray-700"
      >
        <button
          class="px-3 py-1 rounded hover:bg-gray-100 disabled:opacity-50"
          disabled=""
        >
          ← Anterior
        </button>
        <button
          class="px-3 py-1 rounded text-blue-600 underline"
        >
          1
        </button>
        <button
          class="px-3 py-1 rounded hover:bg-gray-100"
        >
          2
        </button>
        <button
          class="px-3 py-1 rounded hover:bg-gray-100"
        >
          3
        </button>
        <button
          class="px-3 py-1 rounded hover:bg-gray-100"
        >
          4
        </button>
        <button
          class="px-3 py-1 rounded hover:bg-gray-100 disabled:opacity-50"
        >
          Próximo →
        </button>
      </nav>
    </div>
    <div
      class="fixed top-0 right-0 w-full max-w-md h-full bg-white shadow-lg transform transition-transform duration-300 ease-in-out z-50 translate-x-full"
    >
      <div
        class="flex items-center justify-between p-4 border-b"
      >
        <h2
          class="text-xl font-semibold"
        >
          Novo Produto
        </h2>
        <button
          class="text-gray-500 hover:text-gray-900 text-xl"
        >
          ×
        </button>
      </div>
      <form
        class="p-6 space-y-4"
      >
        <div>
          <label
            class="block text-sm font-medium mb-1 capitalize"
          >
            nome
          </label>
          <input
            class="w-full border border-gray-300 rounded px-3 py-2"
            name="nome"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1 capitalize"
          >
            categoria
          </label>
          <input
            class="w-full border border-gray-300 rounded px-3 py-2"
            name="categoria"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1 capitalize"
          >
            preco
          </label>
          <input
            class="w-full border border-gray-300 rounded px-3 py-2"
            name="preco"
            type="number"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1 capitalize"
          >
            descricao
          </label>
          <textarea
            class="w-full border border-gray-300 rounded px-3 py-2"
            name="descricao"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
          >
            Imagem do Produto
          </label>
          <div
            class="flex items-center gap-4 mb-2"
          >
            <label
              class="inline-flex items-center gap-2"
            >
              <input
                checked=""
                type="radio"
              />
              Upload
            </label>
            <label
              class="inline-flex items-center gap-2"
            >
              <input
                type="radio"
              />
              URL manual
            </label>
          </div>
          <input
            accept="image/*"
            class="w-full"
            type="file"
          />
        </div>
        <div
          class="flex justify-end gap-2"
        >
          <button
            class="px-4 py-2 bg-gray-200 text-gray-800 rounded hover:bg-gray-300"
            type="button"
          >
            Cancelar
          </button>
          <button
            class="px-4 py-2 bg-blue-600 text-white rounded hover:bg-blue-700"
            type="submit"
          >
            Salvar
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
